http://bugs.gentoo.org/324863

Fix buffer overflow with GCC 4.5 and -U_FORTIFY_SOURCE=2

Patch by Harald van Dijk

--- comm-1.9.2/mozilla/extensions/spellcheck/hunspell/src/hashmgr.cpp
+++ comm-1.9.2/mozilla/extensions/spellcheck/hunspell/src/hashmgr.cpp
@@ -187,7 +187,7 @@
     struct hentry* hp = 
 	(struct hentry *) malloc (sizeof(struct hentry) + wbl + descl);
     if (!hp) return 1;
-    char * hpw = &(hp->word);
+    char * hpw = HENTRY_WORD(hp);
     strcpy(hpw, word);
     if (ignorechars != NULL) {
       if (utf8) {
--- comm-1.9.2/mozilla/extensions/spellcheck/hunspell/src/htypes.hxx
+++ comm-1.9.2/mozilla/extensions/spellcheck/hunspell/src/htypes.hxx
@@ -57,6 +57,8 @@
 #ifndef _HTYPES_HXX_
 #define _HTYPES_HXX_
 
+#include <cstddef>
+
 #define ROTATE_LEN   5
 
 #define ROTATE(v,q) \
@@ -68,7 +70,7 @@
 #define H_OPT_PHON   (1 << 2)
 
 // see also csutil.hxx
-#define HENTRY_WORD(h) &(h->word)
+#define HENTRY_WORD(h) ((char *) h + offsetof(struct hentry, word))
 
 // approx. number  of user defined words
 #define USERWORD 1000
